File: Safrel/Safrel_v1/src/main/java/performancetest/model/VirtualMachine.java
package performancetest.model;


public class VirtualMachine {
    public double Throughput;
    public double ResponseTime_i;
    public double ResponseTime;
    public double NormalizedResponsetime;
    public double Requirement_ResTime; //in ms
    public double Acceptolerance; //in terms of percentage of requirement response time
    public double VM_CPU_g;
    public double VM_Mem_g;
    public double VM_Disk_g;
    public double VM_CPU_i;
    public double VM_Mem_i;
    public double VM_Disk_i;
    public double VM_CPUtil;
    public double VM_MemUtil;
    public double VM_DiskUtil;
    public Double[] VM_SensitivityValues;

    public void CalculateVMThroughput_ResponseTime() {
        double Part1 = (this.VM_CPU_g / this.VM_CPU_i) * this.VM_SensitivityValues[0];
        double Part2 = (this.VM_Mem_g / this.VM_Mem_i) * this.VM_SensitivityValues[1];
        double Part3 = (this.VM_Disk_g / this.VM_Disk_i) * this.VM_SensitivityValues[2];
        double Part4 = this.VM_SensitivityValues[0] + this.VM_SensitivityValues[1] + this.VM_SensitivityValues[2];
        this.Throughput = ((Part1 + Part2 + Part3) / Part4) * 1000.0 / this.ResponseTime_i;
        this.ResponseTime = (double) Math.round((1000.0 / this.Throughput) * 100.0) / 100.0;
    }

    public void CalculateCPUtilImprov() {
        this.VM_CPUtil = this.VM_CPU_i / this.VM_CPU_g;
    }

    public void CalculateMemUtilImprov() {
        this.VM_MemUtil = this.VM_Mem_i / this.VM_Mem_g;
    }

    public void CalculateDiskUtilImprov() {
        this.VM_DiskUtil = this.VM_Disk_i / this.VM_Disk_g;
    }

    public void NormalizeResponseTime() {
        this.NormalizedResponsetime = (2.0 / Math.PI) * Math.atan(this.ResponseTime / this.Requirement_ResTime);
    }

}

File: a2c/src/model/VirtualMachine.py
class VirtualMachine:

    def __init__(self, cpu_g, mem_g, disk_g, cpu_i, mem_i, disk_i, sensitivity_values, response_time_i, Requirement_ResTime, Acceptolerance):
        self.VM_CPU_g = cpu_g
        self.VM_Mem_g = mem_g
        self.VM_Disk_g = disk_g
        self.VM_CPU_i = cpu_i
        self.VM_Mem_i = mem_i
        self.VM_Disk_i = disk_i
        self.VM_SensitivityValues = sensitivity_values
        self.ResponseTime_i = response_time_i
        self.Requirement_ResTime = Requirement_ResTime
        self.Acceptolerance = Acceptolerance
        self.Throughput = 0.0
        self.ResponseTime = 0.0

    def calculate_throughput_response_time(self):
        part1 = (self.VM_CPU_g / self.VM_CPU_i) * self.VM_SensitivityValues[0]
        part2 = (self.VM_Mem_g / self.VM_Mem_i) * self.VM_SensitivityValues[1]
        part3 = (self.VM_Disk_g / self.VM_Disk_i) * \
            self.VM_SensitivityValues[2]
        part4 = sum(self.VM_SensitivityValues)

        # Calculate throughput
        self.Throughput = ((part1 + part2 + part3) / part4) * \
            1000.0 / self.ResponseTime_i

        # Calculate response time
        self.ResponseTime = round((1000.0 / self.Throughput) * 100.0) / 100.0
